### PR TITLE
Fixing Issue #27270 - EFS TypeError

### PR DESCRIPTION
**Pull request by PatrickMurray**
##### SUMMARY
Resolving a TypeError invoked by the AWS EFS module when it attempts to add a Filter type to a List type.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
`lib/ansible/modules/cloud/amazon/efs.py`

##### ANSIBLE VERSION
```
ansible 2.3.1.0
  config file = /home/patrick/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 3.5.2 (default, Nov 17 2016, 17:05:23) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION
See Issue #27270

### CODE REVIEW

**Review by ryansydnor — APPROVED** [review]
:shipit:

File: lib/ansible/modules/cloud/amazon/efs.py [hunk 1]
```diff
@@ -407,8 +407,8 @@ def converge_file_system(self, name, tags, targets):
             changed = filter(
                 lambda sid: not targets_equal(['SubnetId', 'IpAddress', 'NetworkInterfaceId'],
                                               current_targets[sid], targets[sid]), intersection)
-            targets_to_delete = list(targets_to_delete) + changed
-            targets_to_create = list(targets_to_create) + changed
+            targets_to_delete = list(targets_to_delete) + list(changed)
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/efs.py]: Looks good, this fixes the python3 incompatibility traceback. Could also turn those filters into list comprehensions.
I notice that this enables another bug on [line 437](https://github.com/ansible/ansible/pull/27318/files#diff-18b06d7d80bc10ddff5b4aed6edc379cL437). In Python3 that filter item is always found to be True on [line 443](https://github.com/ansible/ansible/pull/27318/files#diff-18b06d7d80bc10ddff5b4aed6edc379cL443) leading to a perpetual changed=true. Would you mind adding a fix there too? 437 could either be wrapped in list() or to make it easier to to understand could also do something like `security_groups_to_update = [sid for sid in intersection if targets[sid].get('SecurityGroups') and current_targets[sid]['SecurityGroups'] != targets[sid]['SecurityGroups']]`. After making that a list you can remove line 443 and just do the iteration over it.

**Review by s-hertel — CHANGES_REQUESTED** [review]

**Comment by abadger** [comment]
Hi @PatrickMurray  are you still interested in working on this but just been busy?  If so, no worries, we're just gearing up for the 2.4.0 release and I'm taking a look at what python3 bugfixes we could include in the release.  If you're not interested, let me know and I'll free up some time to address @s-hertel's requests.

**Comment by abadger** [comment]
Okay, I think I've updated that.  switching to list comprehensions was straight forward.  I did puzzle over what should be done in the check in the second list comprehension and loop around line 437 for a while.  What I finally settled on was keeping the existing check for backwards compatibility.  The comments explain that the existing check has the following behaviour:

* If no SecurityGroups are defined in the targets that are passed in to the module, then the groups are not updated.
* If an empty list is passed in as a target's SecurityGroup, then the code would attempt to clear the target's SecurityGroup.

I'm not 100% sure if that is correct but it is backwards compatible.  There are other ways to construct the check:

* If we didn't want the user to be able to specify an empty list to clear a SecurityGroup:
``` python
security_groups_to_update = [sid for sid in intersection if
+                                         targets[sid].get('SecurityGroups') and
+                                         current_targets[sid]['SecurityGroups'] != targets[sid]['SecurityGroups']]
```

* If we want to force the user to always specify what the SecurityGroups should be (if unspecified, the SecurityGroups would be cleared):
``` python
+            security_groups_to_update = [sid for sid in intersection if
+                                         current_targets[sid]['SecurityGroups'] != targets[sid].get('SecurityGroups', [])]
```

I also left the ```if security_groups_to_update:``` conditional in.  That's so we set ```result = True``` outside of the loop.  May be a pre-mature optimization.

**Comment by abadger** [comment]
ready_for_review

**Comment by s-hertel** [comment]
For creation, if no security group is specified the default is used. Specifiying an empty list results in 
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_o3ttacvy/ansible_module_efs.py", line 642, in <module>
    main()
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_o3ttacvy/ansible_module_efs.py", line 625, in main
    changed = connection.converge_file_system(name=name, tags=tags, targets=targets) or changed
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_o3ttacvy/ansible_module_efs.py", line 454, in converge_file_system
    SecurityGroups=targets[sid].get('SecurityGroups', [])
  File "/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py", line 312, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py", line 601, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.BadRequest: An error occurred (BadRequest) when calling the ModifyMountTargetSecurityGroups operation: Must provide at least one security group.
```
This looks good - maybe the comment to specify an empty list should be removed though.

**Comment by abadger** [comment]
Thanks @s-hertel !  Comment updated.

**Comment by abadger** [comment]
And code updated too

**Review by s-hertel — APPROVED** [review]

**Comment by abadger** [comment]
Merged to devel and cherry-picked to stable-2.3 branch.  Will show up in 2.4.0 and 2.3.3 if we make a 2.3.3 release.

**Pull request merged**